File: providers/aws/src/test/java/cloudspec/aws/ec2/EC2ElasticGpusLoaderTest.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2;

import cloudspec.aws.ec2.loader.EC2ElasticGpuLoader;
import datagen.services.ec2.model.ElasticGpusGenerator;
import org.junit.Test;
import software.amazon.awssdk.services.ec2.model.DescribeElasticGpusResponse;
import software.amazon.awssdk.services.ec2.model.ElasticGpus;

import java.util.function.Consumer;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;

public class EC2ElasticGpusLoaderTest extends EC2LoaderTest {
    private final EC2ElasticGpuLoader loader = new EC2ElasticGpuLoader(clientsProvider);

    @Test
    public void shouldLoadElasticGpus() {
        var elasticGpusList = ElasticGpusGenerator.elasticGpusList();
        when(ec2Client.describeElasticGpus(any(Consumer.class))).thenReturn(
                DescribeElasticGpusResponse.builder()
                                           .elasticGpuSet(elasticGpusList.toArray(new ElasticGpus[0]))
                                           .build()
        );

        var resources = loader.getAll();

        assertNotNull(resources);
        assertEquals(elasticGpusList.size(), resources.size());
        assertTrue(elasticGpusList.containsAll(resources));
    }

    @Test
    public void shouldNotLoadElasticGpusByRandomId() {
        when(ec2Client.describeElasticGpus(any(Consumer.class))).thenReturn(
                DescribeElasticGpusResponse.builder()
                                           .build()
        );

        var resourceOpt = loader.getById(ElasticGpusGenerator.elasticGpuId());

        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isEmpty());
    }

    @Test
    public void shouldLoadElasticGpusById() {
        var elasticGpusList = ElasticGpusGenerator.elasticGpusList(1);
        when(ec2Client.describeElasticGpus(any(Consumer.class))).thenReturn(
                DescribeElasticGpusResponse.builder()
                                           .elasticGpuSet(elasticGpusList.toArray(new ElasticGpus[0]))
                                           .build()
        );

        var elasticGpus = elasticGpusList.get(0);
        var resourceOpt = loader.getById(elasticGpus.elasticGpuId());
        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isPresent());
        assertEquals(resourceOpt.get(), elasticGpus);
    }
}

File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2ElasticGpusLoaderTest.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.core.None
import arrow.core.Some
import datagen.services.ec2.model.ElasticGpusGenerator.elasticGpuId
import datagen.services.ec2.model.ElasticGpusGenerator.elasticGpusList
import io.mockk.every
import org.junit.Test
import software.amazon.awssdk.services.ec2.model.DescribeElasticGpusRequest
import software.amazon.awssdk.services.ec2.model.DescribeElasticGpusResponse
import java.util.function.Consumer
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EC2ElasticGpusLoaderTest : EC2LoaderTest() {
    private val loader = EC2ElasticGpuLoader(clientsProvider)

    @Test
    fun `should load all resources`() {
        val elasticGpusList = elasticGpusList()

        every {
            ec2Client.describeElasticGpus(any<Consumer<DescribeElasticGpusRequest.Builder>>())
        } answers {
            val builder = DescribeElasticGpusRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeElasticGpusRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertTrue(request.elasticGpuIds().isEmpty())

            DescribeElasticGpusResponse.builder()
                    .elasticGpuSet(elasticGpusList)
                    .build()
        }

        val resources = loader.all(emptyList())
        assertNotNull(resources)
        assertEquals(elasticGpusList.size, resources.size)
    }

    @Test
    fun `should not load resource by random id`() {
        every {
            ec2Client.describeElasticGpus(any<Consumer<DescribeElasticGpusRequest.Builder>>())
        } answers {
            val builder = DescribeElasticGpusRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeElasticGpusRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.elasticGpuIds().isEmpty())

            DescribeElasticGpusResponse.builder()
                    .build()
        }

        val resourceOpt = loader.byId(emptyList(), elasticGpuId())
        assertTrue(resourceOpt is None)
    }

    @Test
    fun `should load resource by id`() {
        val elasticGpusList = elasticGpusList()

        every {
            ec2Client.describeElasticGpus(any<Consumer<DescribeElasticGpusRequest.Builder>>())
        } answers {
            val builder = DescribeElasticGpusRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeElasticGpusRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.elasticGpuIds().isEmpty())

            DescribeElasticGpusResponse.builder()
                    .elasticGpuSet(elasticGpusList)
                    .build()
        }

        val elasticGpus = elasticGpusList[0]
        val resourceOpt = loader.byId(emptyList(), elasticGpus.elasticGpuId())
        assertTrue(resourceOpt is Some<*>)
    }
}
